### PR TITLE
Increased Sklearn compatibility

### PR DESCRIPTION
**Pull request by FlorianWilhelm**
In order to fix issue #106 this PR adds a ``BaseEstimator`` Scikit-Learn Mixin if Scikit-Learn is available. This adds ``get_params`` and ``set_params`` which is used by many meta estimators like ``grid_search.RandomizedSearchCV``.

If this PR gets accepted I would also extend the documentation and provide further changes to allow the usage of ``grid_search.RandomizedSearchCV`` for hyper-parameter optimization for LightFM models.

### CODE REVIEW

**Comment by maciejkula** [comment]
Thanks for the changes! Would it be possible to introduce the required
interface changes into the LightFM class without using the sklearn mixins?

That would feel like a cleaner solution to me, and we can enforce it in
tests.

On 2 Sep 2016 3:21 am, "Florian Wilhelm" <notifications@github.com> wrote:

In order to fix issue #106 <https://github.com/lyst/lightfm/issues/106>
this PR adds a BaseEstimator Scikit-Learn Mixin if Scikit-Learn is
available. This adds get_params and set_params which is used by many meta
estimators like grid_search.RandomizedSearchCV.

If this PR gets accepted I would also extend the documentation and provide
further changes to allow the usage of grid_search.RandomizedSearchCV for
hyper-parameter optimization for LightFM models.
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/lyst/lightfm/pull/107
Commit Summary

   - Added get/set_params
   - Some cleanups and easier usage
   - Fixed unit test

File Changes

   - *M* lightfm/__init__.py
   <https://github.com/lyst/lightfm/pull/107/files#diff-0> (5)
   - *A* lightfm/compat_sklearn.py
   <https://github.com/lyst/lightfm/pull/107/files#diff-1> (13)
   - *M* lightfm/lightfm.py
   <https://github.com/lyst/lightfm/pull/107/files#diff-2> (19)
   - *M* tests/test_movielens.py
   <https://github.com/lyst/lightfm/pull/107/files#diff-3> (4)
   - *A* tests/test_sklearn_api.py
   <https://github.com/lyst/lightfm/pull/107/files#diff-4> (6)

Patch Links:

   - https://github.com/lyst/lightfm/pull/107.patch
   - https://github.com/lyst/lightfm/pull/107.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/lyst/lightfm/pull/107>, or mute the thread

**Comment by FlorianWilhelm** [comment]
My first patch was actually to directly implement ``get_params`` and set ``set_params`` in the ``LightFM`` class. Then I looked up the implementation of ``BaseEstimator`` and realized that this is so much smarter since they use introspection in order to collect the necessary attributes of a class that need to bet set and get by the simple convention that ``__init__`` parameters must be equally named to the actual attribute (that's why I renamed ``rng`` to ``random_state``). This way the DRY principle is satisfied and adding new attributes to the class is no longer a source of error since there is no need to add them to an explicitly implemented ``get_params`` and ``set_params`` method. A second way would be to copy over the code of Scikit-learn but this would lead to more than 100 additional lines in the LightFM class and an extra file holding the fallback implementation of ``inspect`` from ``sklearn.externals.funcsigs`` that is needed in case ``signature`` is not available in Python. So this option would pollute LightFM with a lot of lines of code that actually do not belong in there and also contradict the DRY principle.

So in total we have three possible solutions:

 1. Naive and explicit implementation of ``set_params`` and ``get_params``
 2. Smart implementation based on the code of Scikit-Learn
 3. Mixing in the code of Scikit-Learn (like currently done)

The advantage of option 3 also is that people using ``isinstance`` and ``issubclass`` will get the expected results since ``SKLearnLightFM`` really is a ``BaseEstimator``.
I still believe that option 3 is a valid way but I can also change it to option 1 if you want to. I would rather refrain from choosing option 2 since this would introduce a lot of code that needs to be maintained from Scikit-Learn into LightFM.

**Comment by maciejkula** [comment]
I think I'd be happiest with option (1). It may be some duplication but I think it keeps things simple.

Both other options introduce dependencies which I can't control, and, especially with (3), inheritance from a class I don't control. I agree that LightFM should conform to the sklearn interface, but I think inheritance is a very heavy-handed approach to that, especially when we just need a couple of methods.

**Comment by FlorianWilhelm** [comment]
... but it runs on my machine! Nay, just kidding, but there was an IOError, could you just restart the Travis tests? I think that was a temporary issue.

File: tests/test_api.py [hunk 1]
```diff
@@ -278,3 +278,15 @@ def test_predict_ranks():
     # Wrong input dimensions
     with pytest.raises(ValueError):
         model.predict_rank(sp.csr_matrix((5, 5)), num_threads=2)
+
+
+def test_sklearn_api():
```
Comment (maciejkula) [inline tests/test_api.py]: Can you add a rudimentary grid search test just for completeness?

**Comment by maciejkula** [comment]
Looks great with just one minor comment!

**Comment by FlorianWilhelm** [comment]
Well... I am not sure you gonna like this. I got the ``RandomizedSearchCV`` working but actually it is no CV anymore since I am using train_index == test_index. This is necessary due to the fact that with pure collaborative filtering you can't really predict for users and items that are yet unknown. So the training/test split for LightFM is actually accomplished by removing entries from the user-item interaction matrix while Scikit-Learn generates smaller train and test matrices by removing rows. This is a major difference that cannot be dealt with except of rewriting the cross validation generator the way I did. To overcome this I would like to provide a modified RandomizedSearchCV that complies with the requirements of LightFM in another PR. To accomplish this I would copy over all parts I can reuse from ``sklearn.cross_validation`` if this is alright or is Scikit-Learn as dependency a cleaner solution?

**Comment by maciejkula** [comment]
I think this is perfectly sufficient for a test --- thanks for doing this!

**Comment by thisisjl** [comment]
Hi @FlorianWilhelm, @maciejkula, as this issue is old, this might have been implemented and I have not seen it. I've been looking for a way to evaluate hyper-parameters of the LightFM model and also the models in the [Implicit library](https://github.com/benfred/implicit). 
In the same way as you mentioned above, I would like to use sklearn's GridSearchCV or a similar method in combination with a cross validation procedure that creates a training and a testing interaction matrices (such as the [random_train_test_split](https://github.com/lyst/lightfm/blob/2bc601bc7ef571752d698e1d1b6ebdceef6e5427/lightfm/cross_validation.py#L20) function).

Is there any example I could use to understand how to accomplish this? If not, how do you deal with hyper-parameter tuning?

I checked the [test_sklearn_cv](https://github.com/FlorianWilhelm/lightfm/blob/006658f1a2c6cd42d784d6693e462e3ea2c4e207/tests/test_movielens.py#L783), but I do not really understand how I could use it for this purpose.

Thank you

**Comment by maciejkula** [comment]
Cross-validation (as implemented in sklearn) isn't a great fit for collaborative filtering problems, as explained in Florian's comment above.

You can approximate it by doing several independent train-test splits and using those in your hyperparameter search.
